File: 0039-Combination-Sum/lc0039.js
// JavaScript solution.

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    candidates.sort((a,b) => a-b);
    let res = [], ans = [];

    var backtrack = function(cur, residual) {
        if (residual === 0) { ans.push([...res]); }
        else {
            for (let i = cur; i < candidates.length; i++) {
                if (residual >= candidates[i]) {
                    res.push(candidates[i]);
                    backtrack(i, residual-candidates[i]);
                    res.pop();
                }
            }
        }
    };

    backtrack(0, target);
    return ans;
};

File: 0039-Combination-Sum/lc0039.cc
// C++ solution.

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> res;
        vector<vector<int>> ans;
        sort(candidates.begin(), candidates.end());
        backtrack(0, target, candidates, res, ans);
        return ans;
    }

    void backtrack(int cur, int target, vector<int>& candidates, vector<int>& res, vector<vector<int>>& ans) {
        if (!target) {
            ans.push_back(res);
            return;
        }

        for (int i = cur; i < candidates.size(); i++) {
            if (target >= candidates[i]) {
                res.push_back(candidates[i]);
                backtrack(i, target-candidates[i], candidates, res, ans);
                res.pop_back();
            }
        }
    }
};
